AMDGPU/GlobalISel: Add waterfall lowering in regbanklegalize

petar-avramovic opened the pull request:
Add rules for G_AMDGPU_BUFFER_LOAD and implement waterfall lowering
for divergent operands that must be sgpr.

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -117,45 +117,73 @@ static LLT getReadAnyLaneSplitTy(LLT Ty) {
 
-static Register buildReadAnyLane(MachineIRBuilder &B, Register VgprSrc,
-                                 const RegisterBankInfo &RBI);
+typedef std::function<MachineInstrBuilder(MachineIRBuilder &, Register,
```
use `using` instead, also consider using `function_ref` if this is never stored

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -117,45 +117,73 @@ static LLT getReadAnyLaneSplitTy(LLT Ty) {
+  return buildReadLane(
+      B, SgprDst, VgprSrc, RBI,
+      [](MachineIRBuilder &B, Register SgprDst, Register VgprSrc) {
+        return B.buildIntrinsic(Intrinsic::amdgcn_readfirstlane, SgprDst)
```
Not for this PR, but we should really have an opcode for this too instead of having one being an intrinsic and one being a generic opcode

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.h:
```diff
@@ -165,6 +165,8 @@ enum RegBankLLTMappingApplyID {
   Sgpr32Trunc,
 
   // Src only modifiers: waterfalls, extends
+  Sgpr32_W,
+  SgprV4S32_W,
```
Can you add a trailing comment or rename this ? The `_W` suffix is not immediately clear to me

petar-avramovic replied:
Added one above, is it clear now?

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -894,6 +1121,15 @@ void RegBankLegalizeHelper::applyMappingSrc(
+    case Sgpr32_W:
+    case SgprV4S32_W: {
+      assert(Ty == getTyFromID(MethodIDs[i]));
+      if (RB != SgprRB) {
+        SgprWaterfallOperandRegs.insert(Reg);
+      }
```
```suggestion
      if (RB != SgprRB)
        SgprWaterfallOperandRegs.insert(Reg);
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -57,6 +57,226 @@ void RegBankLegalizeHelper::findRuleAndApplyMapping(MachineInstr &MI) {
+        Register CmpReg = MRI.createVirtualRegister(VccRB_S1);
+        B.buildICmp(CmpInst::ICMP_EQ, CmpReg, CurrentLaneParts[i], OpParts[i]);
+
+        if (!CondReg) {
+          CondReg = CmpReg;
+        } else {
+          CondReg = B.buildAnd(VccRB_S1, CondReg, CmpReg).getReg(0);
+        }
```
```suggestion
        if (!CondReg) 
          CondReg = CmpReg;
        else 
          CondReg = B.buildAnd(VccRB_S1, CondReg, CmpReg).getReg(0);
    
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -57,6 +57,226 @@ void RegBankLegalizeHelper::findRuleAndApplyMapping(MachineInstr &MI) {
+
+      // Build the comparison(s), CurrentLaneReg == OpReg.
+      unsigned OpSize = OpTy.getSizeInBits();
+      bool Is64 = OpSize % 64 == 0;
+      unsigned PartSize = Is64 ? 64 : 32;
```
```suggestion
      unsigned PartSize = (OpSize % 64 == 0) ? 64 : 32;
```
I think it's better to inline this one because it's used in one specific case. Personally I'd think it means "IsWave64" at first glance

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -57,6 +57,226 @@ void RegBankLegalizeHelper::findRuleAndApplyMapping(MachineInstr &MI) {
+  const SIRegisterInfo *TRI = ST.getRegisterInfo();
+  const TargetRegisterClass *WaveRC = TRI->getWaveMaskRegClass();
+  unsigned MovExecOpc, MovExecTermOpc, XorTermOpc, AndSaveExecOpc, ExecReg;
+  if (ST.isWave32()) {
```
nit: I think those could go in the class directly so this isn't repeated everytime no ?
The class is instantiated per function anyway

Pierre-vh replied:
So it can be a field, right?

petar-avramovic replied:
Yes, changed to field

Pierre-vh commented on llvm/test/CodeGen/AMDGPU/GlobalISel/regbankselect-amdgcn.raw.buffer.load.ll:
```diff
@@ -1,6 +1,6 @@
 ; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
-; RUN: llc -global-isel -mtriple=amdgcn-mesa-mesa3d -stop-after=regbankselect -regbankselect-fast -o - %s | FileCheck %s
-; RUN: llc -global-isel -mtriple=amdgcn-mesa-mesa3d -stop-after=regbankselect -regbankselect-greedy -o - %s | FileCheck %s
+; RUN: llc -global-isel -new-reg-bank-select -mtriple=amdgcn-mesa-mesa3d -stop-after=amdgpu-regbanklegalize -regbankselect-fast -o - %s | FileCheck %s
```
@arsenm Is it fine to move tests entirely to this new RBSelect, or should we keep coverage for both until the old RB is removed?

arsenm replied:
I guess, how much longer until that happens?

petar-avramovic committed:
AMDGPU/GlobalISel: Add waterfall lowering in regbanklegalize

Add rules for G_AMDGPU_BUFFER_LOAD and implement waterfall lowering
for divergent operands that must be sgpr.

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -203,7 +205,14 @@ class AMDGPURegBankLegalizeCombiner {
-    if (!Src.isVirtual())
+
+    // Skip non-vgpr Dst
+    if ((Dst.isVirtual() && MRI.getRegBankOrNull(Dst) != VgprRB) ||
+        (Dst.isPhysical() && !TRI.isVGPR(MRI, Dst)))
```
```suggestion
    if (Dst.isVirtual() ? (MRI.getRegBankOrNull(Dst) != VgprRB) :  !TRI.isVGPR(MRI, Dst))
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -117,45 +117,72 @@ static LLT getReadAnyLaneSplitTy(LLT Ty) {
-                              Register VgprSrc, const RegisterBankInfo &RBI) {
+static void buildReadLane(MachineIRBuilder &B, Register SgprDst,
+                          Register VgprSrc, const RegisterBankInfo &RBI,
+                          ReadLaneFnTy BuildReadLane) {
```
Make the function a template argument?
